src/rust/engine/engine_pyo3/src/externs/interface/fs.rs

```diff
@@ -211,6 +221,112 @@ impl PyObjectProtocol for PySnapshot {
   }
 }
 
+// -----------------------------------------------------------------------------
+// CreateDigest
+// -----------------------------------------------------------------------------
+
+#[allow(dead_code)]
+struct PyCreateDigest {
+  files: Vec<FileContent>,
+  dirs: Vec<PathBuf>,
+}
+
+impl<'source> FromPyObject<'source> for PyCreateDigest {
```

Comment:
In the medium term, rather than converting from an object defined on the Python side into a rust struct, defining these types the way we do for `PyDigest` would avoid copying from a Python object into a Rust struct by directly creating the relevant types in the constructor.

Since these types are already not being used as Rust types natively, it seems like we could skip directly to doing that here? Basically, define these types similarly to your `pyo3` `PyDigest`.

Then, you'd lift the type with a call to the inverse of yesterday's [`from_pyo3_to_rust_cpython`](https://github.com/pantsbuild/pants/pull/12422#issuecomment-887627859) followed by an extract call:
```
fn create_digest_to_digest(
  context: Context,
  args: Vec<Value>,
) -> BoxFuture<'static, NodeResult<Value>> {
  let create_digest: PyCreateDigest = from_rust_cpython_to_pyo3(&args[0]).extract()?;
  ..
}
```

Reply:
Neat! To be clear, we could do that for PyDigest and PySnapshot too? Define in PyO3 and then call in the current engine/ crate from_rust_cpython_to_pyo3?

For whatever reason, that makes a lot more sense to me than yesterday's from_pyo3_to_rust_cpython discussion.

Reply:
Yes, that was the thrust of yesterday's conversation. The downside I kept mentioning was around typesafety: in this case we start from `Value`, which is a thin wrapper around `rust_cpython`'s `PyObject`. In an actual interface method, we'd be forced to accept `PyObject` and then convert, rather than directly accepting `PyDigest`.

Reply:
I think I like this `from_rust_cpython_to_pyo3` so much because it allows us to start simplifying `nodes.rs` and `intrinsics.rs` by removing parsing code from there. Which will make it easier to see how to migrate the rest of the project.

In terms of crate dependencies and circular imports, sounds like `engine/` can depend on `engine_pyo3/`, but not the other way, right? For now, no need fwict to have a third crate like `engine_pyo3_interface/` which is not a `crate-type = ["cdylib"]`.

Reply:
> For now, no need fwict to have a third crate like engine_pyo3_interface/ which is not a crate-type = ["cdylib"].

Yeah I think this is necessary. With it there, `engine/` can depend on `engine_pyo3/` but won't find any of its types.